coding/python_testing.rst

```diff
@@ -4,6 +4,156 @@ Python Unit Testing
+`pytest`_ provides a much richer execution and reporting environment for tests and can be used to run multiple tests files together.
+`pytest`_ test discovery is much more flexible than that provided by :mod:`unittest` but LSST test files should not take advantage of that flexibility as it can lead to inconsistency in test reports that depend on the specific test runner.
+In particular, care must be taken not to have free functions that use a ``test`` prefix or non-\ :class:`~unittest.TestCase` test classes that are named with a ``Test`` prefix in the test files.
+
+.. _RFC-69: https://jira.lsstcorp.org/browse/RFC-69
+
+`pytest`_ can run tests from more than one file in a single invocation and this can be used to verify that there is no state contaminating later tests.
+To run `pytest`_ use the :command:`py.test` executable:
+
+.. code-block:: shell
+
+   $ py.test tests/*.py
+
+and to ensure that the order of test execution does not matter it is useful to sometimes run the tests in reverse order:
+
+.. code-block:: shell
+
+   $ py.test `ls -r tests/*.py`
+
+When writing tests it is important that tests are skipped using the proper :mod:`unittest` :ref:`skipping framework <python:unittest-skipping>` rather than returning from the test early.
+:mod:`unittest` can support skipping of individual tests and entire classes using decorators, and exceptions can also be used.
+LSST code sometimes raises skip exceptions in :meth:`~unittest.TestCase.setUp` or :meth:`~unittest.TestCase.setUpClass` class methods and :mod:`unittest` provides much flexibility in how skipping can be implemented.
+It is also possible to indicate that a particular test is expected to fail, to the extent that it is reported it a test unexpectedly passes.
+Expected failures can be used to write test code that triggers a reported bug before the fix to the bug has been implemented and without causing the continuous integration system to die.
+One of the advantages of using a modern test runner such as `pytest`_ is that it is very easy to generate machine-readable pass/fail/skip/xfail statistics to see how the system is evolving over time.
+
+Using a shared base class
+=========================
+
+For some tests it is helpful to provide a base class and then share it amongst multiple test classes that are configured with different attributes.
+If this is required be careful to not have helper functions prefixed with ``test`` and do not have the base class named with a ``Test`` prefix and ensure it does not inherit from :class:`~unittest.TestCase`, if you do `pytest`_ will attempt to find tests inside it.
+Historically LSST code has dealt with this by creating a test suite that only includes the classes to be tested, omitting the base class.
+This can not work in a `pytest`_ environment, so instead multiple inheritance should be used such that the test classes inherit from :class:`~unittest.TestCase` and the support base class.
```

Comment:
I think it would suffice to say "omitting the base class. This does not work in a pytest environment". You already explained what to do instead.